### PR TITLE
Make entity selection and collision boxes independently settable

### PR DESCRIPTION
**Pull request by stujones11**
Addresses #6199

This PR adds an optional `selectionbox` property to entities that overrides the collision box dimensions, otherwise used as the default. This is particularly useful for vehicle mods where the collision box often needs to be a different size or shape to the vehicle itself.

~~Note that 'zeroing' the selection box table will cause the visible selection box to default to the collision box dimensions, however,~~ a `pointable` property has also been included as a means to prevent selection.

This PR will also make entities more consistent with nodes in terms of available properties, IMO these ones in particular are even more important for entities.

### CODE REVIEW

File: src/content_cao.cpp [hunk 1]
```diff
@@ -1501,7 +1501,14 @@ void GenericCAO::processMessage(const std::string &data)
 	if (cmd == GENERIC_CMD_SET_PROPERTIES) {
 		m_prop = gob_read_set_properties(is);
 
-		m_selection_box = m_prop.collisionbox;
+		aabb3f no_selection = aabb3f(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f);
```
Comment (SmallJoker) [inline src/content_cao.cpp]: This is an ugly workaround. Rather extend `ClientEnvironment::getSelectedActiveObject` to skip the objects that aren't pointable.

File: doc/lua_api.txt [hunk 2]
```diff
@@ -3956,6 +3956,8 @@ Definition tables
         collide_with_objects = true, -- collide with other objects if physical = true
         weight = 5,
         collisionbox = {-0.5, 0.0, -0.5, 0.5, 1.0, 0.5},
+        selectionbox = {0, 0, 0, 0, 0, 0}, -- defaults to collision box dimensions
```
Comment (SmallJoker) [inline doc/lua_api.txt]: If it defaults to the collision box, then why is this table different from the other?

File: src/content_cao.cpp (hunk 1 above)
Comment (stujones11) [inline src/content_cao.cpp]: Okay, I will look at that, however, the 'zeroed' aabb box also denotes that no selection box has been specified. I did say that it wasn't ideal but I see no other way of knowing whether a selection box was set in the entity def.

File: doc/lua_api.txt (hunk 2 above)
Comment (stujones11) [inline doc/lua_api.txt]: Because this _is_ the default setting and what you will see if you dump the property table.

File: src/content_cao.cpp (hunk 1 above)
Comment (stujones11) [inline src/content_cao.cpp]: Rather than make changes to `ClientEnvironment::getSelectedActiveObject` I think it would be more better to include the `pointable` check in `GenericCAO::getSelectionBox` where it already returns `false` for invisible objects and consequently skipped by `getSelectedActiveObject`

**Comment by stujones11** [comment]
@SmallJoker Cheers for the prompt review, I think I have now addressed the issues you raised. This PR is really quite trivial and it would be nice to see it merged before I start getting rebase conflicts ;-)

File: src/content_cao.cpp [hunk 3]
```diff
@@ -614,7 +614,7 @@ GenericCAO::~GenericCAO()
 bool GenericCAO::getSelectionBox(aabb3f *toset) const
 {
 	if (!m_prop.is_visible || !m_is_visible || m_is_local_player
-			|| getParent() != NULL){
+			|| !m_prop.pointable || getParent() != nullptr) {
```
Comment (SmallJoker) [inline src/content_cao.cpp]: `getParent()` does not return `nullptr`. Either revert this change to `NULL` or correct the function definitions and all lines where that function is checked against returning `NULL`. Never, ever mix these two.

File: src/object_properties.cpp [hunk 4]
```diff
@@ -71,6 +73,9 @@ void ObjectProperties::serialize(std::ostream &os) const
 	writeF1000(os, weight);
 	writeV3F1000(os, collisionbox.MinEdge);
 	writeV3F1000(os, collisionbox.MaxEdge);
+	writeV3F1000(os, selectionbox.MinEdge);
+	writeV3F1000(os, selectionbox.MaxEdge);
+	writeU8(os, pointable);
```
Comment (SmallJoker) [inline src/object_properties.cpp]: needs version bump for backwards compatibility.

File: src/content_cao.cpp (hunk 3 above)
Comment (Dumbeldor) [inline src/content_cao.cpp]: I think it's better to keep the nullptr and modify getParent().

Comment (Dumbeldor) [inline src/content_cao.cpp]: I think it's better to keep the nullptr and modify getParent().

Comment (stujones11) [inline src/content_cao.cpp]: Okay, I will revert that last commit asap, seems like the safest option.

File: src/object_properties.cpp (hunk 4 above)
Comment (stujones11) [inline src/object_properties.cpp]: Should this be done as part of this PR? I have seen other new entity properties added without the need for that. I'might not too sure how to go about that, TBH.

Comment (SmallJoker) [inline src/object_properties.cpp]: This is only possible by adding the new values in this output-stream below the already existing ones. Otherwise, older clients will read garbage from the new "gap" you opened here.

Comment (stujones11) [inline src/object_properties.cpp]: Okay, that makes sense, are there any examples of how to bump the version, another PR perhaps? Or do you think it would be better just to move the new fields to the end?

Comment (SmallJoker) [inline src/object_properties.cpp]: KISS. Move the fields to the end.

**Comment by paramat** [comment]
This will be useful for my car mod, currently the selectionbox is set to the collisionbox which by necessity is buried inside the vehicle and barely visible.

**Comment by stujones11** [comment]
I have made the requested changes and squashed the commits but am unable to actually test this right now. Will do this asap but given the changes, I think it should be fine.

**Comment by SmallJoker** [comment]
Tested using `pointable = false` (and `= true` in minetest_game/mods/player_api/api.lua L49.
Works fine, but only when the server uses this commit. For some reason there's no SerializationError in this case when reading outside the actual packet. As a result of this, `pointable` is always set to `false` when playing on older servers. I.e. none of the entities is selectable.

**Comment by stujones11** [comment]
@SmallJoker Thanks for testing, I have done as you suggested and inverted the case. However, I am wondering whether backwards compatibility is really that important right now. I mean the current development builds are pretty much broken on older servers anyway.

**Comment by stujones11** [comment]
@SmallJoker I noticed a slight inconsistency with this on the server-side while looking at the ray-casting issue, I hope you are still okay with this.

**Comment by SmallJoker** [comment]
@stujones11 I don't see any regression - If you didn't find any problems while testing either, I'm fine fine with the new change.

**Comment by stujones11** [comment]
Thanks but I spotted a problem, the server-side properties are updated first therefore it's no use updating the `m_prop` client-side, the server does not see this. I have now moved the selection box default to the script api, this is actually much better since it avoids the need to compare `aabb3f` boxes, which I never did like doing.

**Comment by stujones11** [comment]
Reworded documentation since now the selection box will be the same as the collision box if no selection box was specified in the entity definition. A 'zeroed' selection box will now behave as expected but `pointable = false` should still be used to prevent selection.

**Comment by stujones11** [comment]
I have now extended these properties to also work with players, however, this requires some slight modifications to player_api to make them actually useable.

Note this last commit also incorporates the ray-casting fix I made here #6241 so that would no longer be necessary if this were to be merged first.

File: src/object_properties.cpp [hunk 5]
```diff
@@ -98,6 +100,9 @@ void ObjectProperties::serialize(std::ostream &os) const
 	writeF1000(os, automatic_face_movement_max_rotation_per_sec);
 	os << serializeString(infotext);
 	os << serializeString(wield_item);
+	writeV3F1000(os, selectionbox.MinEdge);
+	writeV3F1000(os, selectionbox.MaxEdge);
+	writeU8(os, !pointable);
```
Comment (nerzhul) [inline src/object_properties.cpp]: why invert the boolean ? please keep the right value

Comment (stujones11) [inline src/object_properties.cpp]: This was requested by @SmallJoker from the emote, it seems @sfan5 agrees.

Comment (sfan5) [inline src/object_properties.cpp]: since `readU8` defaults to false if out-of-bounds this needs to be inverted to get pointable to be `true` if omitted

Comment (nerzhul) [inline src/object_properties.cpp]: @sfan5 it's crazy and not very dev friendly, we should fix this later... i dismiss my request

Comment (sfan5) [inline src/object_properties.cpp]: indeed this should be fixed

Comment (stujones11) [inline src/object_properties.cpp]: This can be fixed later without breaking anything so is there any chance one of you could give a second approval on this? It's fairly trivial stuff but useful for modders.

Comment (paramat) [inline src/object_properties.cpp]: Backwards compatibility is no concern at the moment as we have just merged 2 compatibility breaking commits, so just choose whatever is best without considering new-old server-client compatibility.

Comment (paramat) [inline src/object_properties.cpp]: .. which i guess is not inverting.

**Comment by stujones11** [comment]
Rebased and squashed into 2 commits. I have reverted the backwards compatibility hack but left the revert commit so the merging developer can decide whether to include it or not.
